### PR TITLE
2.0: Remove SuccessWithValue

### PR DESCRIPTION
**Pull request by bradjc**
### Pull Request Overview

This pull request removes `SuccessWithValue` for 2.0. Other than IPC, nothing was using SuccessWithValue anymore.

### Testing Strategy

travis

### TODO or Help Wanted

Need the IPC updates first.

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

**Review by hudson-ayers — APPROVED** [review]
Looks good modulo the IPC change, of course

File: kernel/src/returncode.rs [hunk 1]
```diff
@@ -42,7 +40,6 @@ pub enum ReturnCode {
 impl From<ReturnCode> for isize {
     fn from(original: ReturnCode) -> isize {
         match original {
-            ReturnCode::SuccessWithValue { value } => value as isize,
```
Comment (phil-levis) [inline kernel/src/returncode.rs:45]: HURRAY!

**Comment by bradjc** [comment]
Ok new plan...we leave `SuccessWithValue` but vow to never use it.

But seriously, this change seems to increase the imix kernel size by 4512 bytes.

Before:
```
make
   Compiling kernel v0.1.0 (/Users/bradjc/git/tock/kernel)
   Compiling cortexm v0.1.0 (/Users/bradjc/git/tock/arch/cortex-m)
   Compiling capsules v0.1.0 (/Users/bradjc/git/tock/capsules)
   Compiling cortexm4 v0.1.0 (/Users/bradjc/git/tock/arch/cortex-m4)
   Compiling sam4l v0.1.0 (/Users/bradjc/git/tock/chips/sam4l)
   Compiling components v0.1.0 (/Users/bradjc/git/tock/boards/components)
   Compiling imix v0.1.0 (/Users/bradjc/git/tock/boards/imix)
    Finished release [optimized + debuginfo] target(s) in 41.70s
   text	   data	    bss	    dec	    hex	filename
 188373	     64	  32704	 221141	  35fd5	/Users/bradjc/git/tock/target/thumbv7em-none-eabi/release/imix
dd516105645d93567b70c598c75fe2e2cb3d4de9b83817aa4f4b49bdbd88746e	/Users/bradjc/git/tock/target/thumbv7em-none-eabi/release/imix.bin
```
after
```
make
   Compiling imix v0.1.0 (/Users/bradjc/git/tock/boards/imix)
    Finished release [optimized + debuginfo] target(s) in 18.77s
   text	   data	    bss	    dec	    hex	filename
 192885	     64	  32704	 225653	  37175	/Users/bradjc/git/tock/target/thumbv7em-none-eabi/release/imix
3ba74f8fd1b72bdbb0c9dddf57609caf2ce8834b561b12eda0aebd07c589f592	/Users/bradjc/git/tock/target/thumbv7em-none-eabi/release/imix.bin
```

**Comment by bradjc** [comment]
Just adding the `SuccessWithValue` enum value back "fixes" the issue and reduces the size relative to the 2.0 branch:

```
make
   Compiling kernel v0.1.0 (/Users/bradjc/git/tock/kernel)
   Compiling cortexm v0.1.0 (/Users/bradjc/git/tock/arch/cortex-m)
   Compiling capsules v0.1.0 (/Users/bradjc/git/tock/capsules)
   Compiling cortexm4 v0.1.0 (/Users/bradjc/git/tock/arch/cortex-m4)
   Compiling sam4l v0.1.0 (/Users/bradjc/git/tock/chips/sam4l)
   Compiling components v0.1.0 (/Users/bradjc/git/tock/boards/components)
   Compiling imix v0.1.0 (/Users/bradjc/git/tock/boards/imix)
    Finished release [optimized + debuginfo] target(s) in 34.42s
   text	   data	    bss	    dec	    hex	filename
 188309	     64	  32704	 221077	  35f95	/Users/bradjc/git/tock/target/thumbv7em-none-eabi/release/imix
3e6578da0741e58e8f0b802d445558df4c1394f8f898dd064ea0d39bde04f7ae	/Users/bradjc/git/tock/target/thumbv7em-none-eabi/release/imix.bin
```

**Comment by phil-levis** [comment]
I will look into this. I'll try to improve the print tool to see what's happening.

**Comment by hudson-ayers** [comment]
Rebased ontop of 2.0-dev now that I merged master in

**Review by hudson-ayers — APPROVED** [review]

**Comment by hudson-ayers** [comment]
After the rebase, this PR adds just under 2kB of code, but Imix now fits.

**Comment by hudson-ayers** [comment]
Some tidbits from my investigating the size changes:

- Increase seems spread across many functions
- Marking all returnCode functions inline(never) → small improvement
- Marking inline(always) → little worse
- mem::size_of(Returncode now): 1 byte. Before it was 8 bytes!
```rust
pub enum ReturnCode {
    SuccessWithValue { value },
    SUCCESS
```
Total Imix size: 184996 bytes
```rust
pub enum ReturnCode {
    SUCCESS,
```
Total Imix size: 186980
```rust
#[repr(isize)]
pub enum ReturnCode {
    SUCCESS,
```
Total Imix size: 184485
```rust
#[repr(rust)]
pub enum ReturnCode {
    SUCCESS,
// ...
#[repr(rust)] // previously was #[repr(usize)]
pub enum ErrorCode {
    SUCCESS,
```
Total Imix size: 185541

Hypothesis: Making ReturnCode 1 byte instead of 4 decreases stack memory use, however doing so also increases code size, because now kernel has to sign extend values when moving ReturnCodes in/out of a word-size register at the userspace boundary.

**Comment by alevy** [comment]
Following some side discussion, @hudson-ayers's theory seems to hold water and his proposal to just force `ReturnCode` to `usize`/`isize` seems strictly positive---there is no semantic difference within the kernel and it ends up resulting in an overall improvement in code space for this PR.

A note: I'm very skeptical that we get any actual memory savings from the ReturnCode being smaller than a word since likely the compiler is using a word for it on the stack anyway in most cases.

**Review by alevy — APPROVED** [review]

**Review by ppannuto — APPROVED** [review]
bors r+

**Review by lschuermann — APPROVED** [review]

**Comment by phil-levis** [comment]
> A note: I'm very skeptical that we get any actual memory savings from the ReturnCode being smaller than a word since likely the compiler is using a word for it on the stack anyway in most cases.

This seems to be a case where you can trade off RAM for flash; given we are much more cramped for flash, a possible few extra bytes of RAM here or there are well worth it.

**Pull request merged**